pyteal/ast/router_test.py

```diff
@@ -0,0 +1,533 @@
+        safe_clear_state_delete,
+        pt.Approve(),
+        pt.Log(pt.Bytes("message")),
+    ]
+    for bare_call in BARE_CALL_CASES:
+        wrapped: pt.Expr = pt.ASTBuilder.wrap_handler(False, bare_call)
+        match bare_call:
+            case pt.Expr():
+                if bare_call.has_return():
+                    assert wrapped == bare_call
+                else:
+                    assert wrapped == pt.Seq(bare_call, pt.Approve())
+            case pt.SubroutineFnWrapper() | pt.ABIReturnSubroutine():
+                assert wrapped == pt.Seq(bare_call(), pt.Approve())
+            case _:
+                raise pt.TealInputError("how you got here?")
+
+    ERROR_CASES = [
+        (
+            pt.Int(1),
+            f"bare appcall handler should be TealType.none not {pt.TealType.uint64}.",
+        ),
+        (
+            returning_u64,
+            f"subroutine call should be returning TealType.none not {pt.TealType.uint64}.",
+        ),
+        (
+            mult_over_u64_and_log,
+            "subroutine call should take 0 arg for bare-app call. this subroutine takes 2.",
+        ),
+        (
+            eine_constant,
+            f"abi-returning subroutine call should be returning void not {pt.abi.Uint64TypeSpec()}.",
+        ),
+        (
+            take_abi_and_log,
+            "abi-returning subroutine call should take 0 arg for bare-app call. this abi-returning subroutine takes 1.",
+        ),
+        (
+            1,
+            "bare appcall can only accept: none type Expr, or Subroutine/ABIReturnSubroutine with none return and no arg",
+        ),
+    ]
+    for error_case, error_msg in ERROR_CASES:
+        with pytest.raises(pt.TealInputError) as bug:
+            pt.ASTBuilder.wrap_handler(False, error_case)
+        assert error_msg in str(bug)
+
+
+def test_wrap_handler_method_call():
+    with pytest.raises(pt.TealInputError) as bug:
+        pt.ASTBuilder.wrap_handler(True, not_registrable)
+    assert "method call ABIReturnSubroutine is not routable" in str(bug)
+
+    with pytest.raises(pt.TealInputError) as bug:
+        pt.ASTBuilder.wrap_handler(True, safe_clear_state_delete)
+    assert "method call should be only registering ABIReturnSubroutine" in str(bug)
+
+    ONLY_ABI_SUBROUTINE_CASES = list(
+        filter(lambda x: isinstance(x, pt.ABIReturnSubroutine), GOOD_SUBROUTINE_CASES)
+    )
+    for abi_subroutine in ONLY_ABI_SUBROUTINE_CASES:
+        wrapped: pt.Expr = pt.ASTBuilder.wrap_handler(True, abi_subroutine)
+        assembled_wrapped: pt.TealBlock = assemble_helper(wrapped)
+
+        args: list[pt.abi.BaseType] = [
+            spec.new_instance()
+            for spec in typing.cast(
+                list[pt.abi.TypeSpec], abi_subroutine.subroutine.expected_arg_types
+            )
+        ]
+
+        loading: list[pt.Expr]
+
+        if abi_subroutine.subroutine.argument_count() > pt.METHOD_ARG_NUM_LIMIT:
+            sdk_last_arg = pt.abi.TupleTypeSpec(
```

Comment:
This test for > 15 arguments looks like it copies the implementation code it's responsible for testing. This is generally not a great idea, so instead of/in addition to this, could you manually construct a test for this case?

Reply:
Blackbox testing is always appreciated, but I wanted to have more confidence in our code generation, so I made #375